File: java/06-io-operations/IoOperations.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.List;

public class IoOperations {
    public static void main(String[] args) throws IOException {
        // 1. Formatted console output
        String name = "Grace";
        int commits = 42;
        double coverage = 0.8756;
        System.out.printf("%s made %d commits (%.1f%% coverage).%n", name, commits, coverage * 100);

        // 2. Write a list of lines to a file (creates or overwrites it)
        Path log = Path.of("build.log");
        List<String> entries = List.of("compile: OK", "test: OK", "package: OK");
        Files.write(log, entries);
        System.out.println("Wrote " + entries.size() + " lines to " + log.getFileName());

        // 3. Read the whole file back into memory
        List<String> lines = Files.readAllLines(log);
        System.out.println("File contents:");
        for (int i = 0; i < lines.size(); i++) {
            System.out.printf("  line %d -> %s%n", i + 1, lines.get(i));
        }

        // 4. Append one more line without erasing the file
        Files.writeString(log, "deploy: OK" + System.lineSeparator(), StandardOpenOption.APPEND);
        System.out.println("After append, total lines: " + Files.readAllLines(log).size());

        // 5. Handle I/O errors with try-catch instead of crashing
        try {
            Files.readAllLines(Path.of("does-not-exist.txt"));
        } catch (NoSuchFileException e) {
            System.out.println("Handled missing file: " + e.getFile());
        }
    }
}

File: kotlin/06-io-operations/IoOperations.kt
import java.io.File

fun main() {
    // print() has no newline; println() adds one
    print("No newline here. ")
    println("This ends the line.")

    // Values concatenate directly into strings
    println("Numbers: " + 42 + ", " + 3.14)

    // Formatted output with printf-style specifiers
    val pi = 3.14159265
    println("Pi to 2 decimals: %.2f".format(pi))
    println("Padded: %5d|".format(42))       // right-aligned in 5 columns
    println("Name: %-10s|".format("Kotlin"))  // left-aligned in 10 columns

    // Write text to a file (creates or overwrites)
    val file = File("greetings.txt")
    file.writeText("Hello from Kotlin\n")
    file.appendText("Second line\n")

    // Read the whole file back as a single string
    val content = file.readText()
    print(content)

    // Read the file as a list of lines and number them
    println("--- lines ---")
    file.readLines().forEachIndexed { index, line ->
        println("${index + 1}: $line")
    }
}
